src/Abstractions/TaskFailureDetails.cs

```diff
@@ -35,12 +36,37 @@ public override string ToString()
     /// </remarks>
     /// <typeparam name="T">The type of exception to test against.</typeparam>
     /// <returns>
-    /// Returns <c>true</c> if the <see cref="ErrorType"/> value matches <typeparamref name="T"/>; <c>false</c> otherwise.
+    /// <c>true</c> if the <see cref="ErrorType"/> value matches <typeparamref name="T"/>; <c>false</c> otherwise.
     /// </returns>
     public bool IsCausedBy<T>() where T : Exception
     {
+        return this.IsCausedBy(typeof(T));
+    }
+
+    /// <returns>
+    /// <c>true</c> if the <see cref="ErrorType"/> value matches <paramref name="type"/>; <c>false</c> otherwise.
+    /// </returns>
+    /// <exception cref="ArgumentException">If <paramref name="type"/> is not an exception type.</exception>
+    /// <inheritdoc cref="IsCausedBy{T}"/>
+    public bool IsCausedBy(Type type)
+    {
+        if (!typeof(Exception).IsAssignableFrom(type))
+        {
+            throw new ArgumentException($"The type {type} is not an exception type.", nameof(type));
+        }
+
+        // This check works for core .NET exception types
         this.exceptionType ??= Type.GetType(this.ErrorType, throwOnError: false);
-        return this.exceptionType != null && typeof(T).IsAssignableFrom(this.exceptionType);
+
+        // This check works for custom exception types defined in the app's assembly
+        this.exceptionType ??= Assembly.GetCallingAssembly().GetType(this.ErrorType);
+
+        // This last check works for exception types defined in any loaded assembly (e.g. NuGet packages, etc.)
+        this.exceptionType ??= AppDomain.CurrentDomain.GetAssemblies()
+            .Select(a => a.GetType(this.ErrorType, throwOnError: false))
```

Comment:
How will this behave if two assemblies have an exception with the same full name? Will we get false positives?

I wonder if instead we should be using `Type.AssemblyQualifiedName`, which would include assembly and public key. I think `Type.GetType(..)` with that value will then work.

Reply:
What you're suggesting (or some variant of it) would be ideal, but it's unfortunately a breaking change, which is why I didn't go that route. Yes, there's a possibility of false positives, but I think it's extremely unlikely, to the point where I wouldn't block this PR for that. From my view, no app should ever have the same exception type name + namespace in two different assemblies with the expectation that they be treated differently. In many cases you wouldn't even be able to compile such code.

Reply:
`AssemblyQualifiedName` includes versions, which might cause an issue if an app is redeployed with a higher dependency. Local testing shows `Type.GetType` works with just type name + assembly name:

``` CSharp
Type t1 = Type.GetType("System.Text.Json.JsonException"); // t1 is null
Type t2 = Type.GetType("System.Text.Json.JsonException, System.Text.Json"); // t2 is not null
```

Reply:
> What you're suggesting (or some variant of it) would be ideal, but it's unfortunately a breaking change, which is why I didn't go that route. Yes, there's a possibility of false positives, but I think it's extremely unlikely, to the point where I wouldn't block this PR for that. From my view, no app should ever have the same exception type name + namespace in two different assemblies with the expectation that they be treated differently. In many cases you wouldn't even be able to compile such code.

Would it be a breaking change? We aren't adding any extra fields, just appending it to `ErrorType` field. But it is still backwards compatible: no assembly being there still works. So both of these will 'work':

`ErrorType == "System.Text.Json.JsonException"` -> `IsCausedBy<T>()` continues to not work.
`ErrorType == "System.Text.Json.JsonException, System.Text.Json"` -> `IsCausedBy<T>()` now works.

Reply:
Part of me also worries a bit about the implications of trying to load a type using a specified assembly name. My biggest fear is that we suddenly expose ourselves to an assembly versioning issue where, for example, we save v1 of an assembly name into the history, but then the app is later updated to use v2 of that assembly. In such cases, I expect we would no longer be able to load that exception type, causing the orchestration to go down an unexpected path resulting in a non-determinism exception (A paranoid part of me also wonders if there could be negative security implications of loading arbitrary assemblies into memory...).

Reply:
For your two points:

1. If that type is now gone from the assembly, that assembly has had a breaking change. I don't think we can really prevent anything regarding that. It is not a new or unique issue either - any polymorphic deserialization has the same issue.
2. We won't be loading arbitrary assemblies into memory. This would load only assemblies that are known to the application (I believe it will be limited to assemblies in the deps.json).

Reply:
Just out of curiosity - 
@jviau are you saying we'd call "Type.AssemblyQualifiedName" and perform a string manipulation to remove the assembly version and everything that comes after it?

So to go from:

> Type.AssemblyQualifiedName: System.Text.Json.JsonException, System.Text.Json, Version=8.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51

to:

> "System.Text.Json.JsonException, System.Text.Json"

perhaps with a "split by delimeter" operation? Would be nice if no string manipulation was needed at all.

Reply:
I think `$"{t.FullName}, {t.Assembly.GetName().Name}"` will be sufficient.